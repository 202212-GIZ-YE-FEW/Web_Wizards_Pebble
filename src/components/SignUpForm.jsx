import { useRouter } from "next/navigation";

import { useAuthContext } from "@/context/AuthContext";

import FormFooter from "./FormFooter";
import InputForm from "./InputForm";
import SignButton from "./SignButton";
import SignOptions from "./SignOptions";
import SignTitle from "./SignTitle";
// TODO: use dynamic path once testing doesnt conflicts paths.
import { signUp } from "../../lib/useAuth";

const SignUpForm = ({ t }) => {
    const router = useRouter();
    const { user } = useAuthContext();
    /**
     *
     * @param {import('react').SyntheticEvent} e
     */
    const handleSubmit = async (e) => {
        e.preventDefault();
        if (user) return;
        const formData = new FormData(e.target);
        const firstName = formData.get("name");
        const lastName = formData.get("surname");
        const email = formData.get("email");
        const password = formData.get("password");
        try {
            await signUp(email, password, { firstName, lastName });
            router.push("/");
        } catch (error) {
            alert(error);
        }
    };
    return (
        <>
            <div className='w-full lg:w-3/5 flex-grow-1 relative px-4 lg:px-8 order-1'>
                <div className='mx-auto lg:max-w-lg md:max-w-3xl py-4 lg:py-8 flex flex-col'>
                    <SignTitle text={t("signUp")} />
                    <SignOptions t={t} />
                    <form className='-mx-3 order-3' onSubmit={handleSubmit}>
                        <div className='sm:flex flex-row hidden'>
                            <InputForm
                                placeholder={t("name")}
                                label={t("name")}
                                name='name'
                                type='text'
                                hasFloatingLabel
                                isRequired
                                id='name'
                            />
                            <InputForm
                                placeholder={t("surname")}
                                label={t("surname")}
                                name='surname'
                                type='text'
                                hasFloatingLabel
                                isRequired
                                id='surname'
                            />
                        </div>
                        <InputForm
                            placeholder={t("email")}
                            label={t("email")}
                            name='email'
                            type='email'
                            hasFloatingLabel
                            isRequired
                            id='email'
                        />
                        <InputForm
                            placeholder={t("password")}
                            label={t("password")}
                            name='password'
                            type='password'
                            hasFloatingLabel
                            togglePasswordLabel='Show/Hide password'
                            togglePassword='true'
                            isRequired
                            id='password'
                            passwordCriteria={t("passwordCriteria")}
                        />
                        <FormFooter
                            message={t("haveAccount")}
                            linkHref='/signin'
                            linkText={t("signIn")}
                        />
                        <SignButton text={t("signUp")} />
                    </form>
                </div>
            </div>
        </>
    );
};

export default SignUpForm;
