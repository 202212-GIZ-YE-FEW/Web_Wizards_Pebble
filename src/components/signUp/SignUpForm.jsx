import { useRouter } from "next/navigation";
import { useEffect, useState } from "react";

import FormFooter from "@/components/shared/FormFooter";
import InputForm from "@/components/shared/InputForm";
import SignButton from "@/components/shared/SignButton";
import SignOptions from "@/components/shared/SignOptions";
import SignTitle from "@/components/shared/SignTitle";

import { useAuthContext } from "@/context/AuthContext";

import AuthErrorBox from "../shared/AuthErrorBox";
// TODO: use dynamic path once testing doesnt conflicts paths.
import { signUp } from "../../../lib/useAuth";

const SignUpForm = ({ t }) => {
    const router = useRouter();
    const { user } = useAuthContext();
    const [errors, setErrors] = useState([]);
    useEffect(() => {
        if (user && user.emailVerified) router.push("/");
    }, [user, router]);
    const [loading, setLoading] = useState(false);
    /**
     *
     * @param {import('react').SyntheticEvent} e
     */
    const handleSubmit = async (e) => {
        e.preventDefault();
        if (user && user.emailVerified) return;
        const formData = new FormData(e.target);
        const firstName = formData.get("name");
        const lastName = formData.get("surname");
        const email = formData.get("email");
        const password = formData.get("password");
        try {
            setLoading(true);
            await signUp(email, password, { firstName, lastName });
        } catch (error) {
            setLoading(false);
            setErrors(error.code || "auth/unknown");
            return;
        }
        setLoading(false);
        router.push("/");
    };
    return (
        <>
            <div className='w-full lg:w-3/5 flex-grow-1 relative px-4 lg:px-8 order-1'>
                <div className='mx-auto lg:max-w-lg md:max-w-3xl py-4 lg:py-8 flex flex-col'>
                    <SignTitle text={t("signUp")} />
                    <SignOptions t={t} />
                    {errors.length > 0 && (
                        <AuthErrorBox t={t} errors={errors} />
                    )}
                    <form className='-mx-3 order-3' onSubmit={handleSubmit}>
                        <div className='sm:flex flex-row hidden'>
                            <InputForm
                                autoFocus
                                placeholder={t("name")}
                                label={t("name")}
                                name='name'
                                type='text'
                                hasFloatingLabel
                                isRequired
                                id='name'
                            />
                            <InputForm
                                placeholder={t("surname")}
                                label={t("surname")}
                                name='surname'
                                type='text'
                                hasFloatingLabel
                                isRequired
                                id='surname'
                            />
                        </div>
                        <InputForm
                            placeholder={t("email")}
                            label={t("email")}
                            name='email'
                            type='email'
                            hasFloatingLabel
                            isRequired
                            id='email'
                        />
                        <InputForm
                            placeholder={t("password")}
                            label={t("password")}
                            name='password'
                            type='password'
                            hasFloatingLabel
                            togglePasswordLabel='Show/Hide password'
                            togglePassword='true'
                            isRequired
                            id='password'
                            passwordCriteria={t("passwordCriteria")}
                        />
                        <FormFooter
                            message={t("haveAccount")}
                            linkHref='/signin'
                            linkText={t("signIn")}
                        />
                        <SignButton text={t("signUp")} loading={loading} />
                    </form>
                </div>
            </div>
        </>
    );
};

export default SignUpForm;
