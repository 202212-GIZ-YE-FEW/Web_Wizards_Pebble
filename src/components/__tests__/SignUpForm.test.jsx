import { I18nContext } from "next-i18next";
import renderer from "react-test-renderer";

import SignUpForm from "../SignUpForm";

it("renders correctly", () => {
    const i18n = {
        language: "en",
        t: (str) => str,
    };
    const tree = renderer
        .create(
            <I18nContext.Provider value={i18n}>
                <SignUpForm t={i18n.t} />
            </I18nContext.Provider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
