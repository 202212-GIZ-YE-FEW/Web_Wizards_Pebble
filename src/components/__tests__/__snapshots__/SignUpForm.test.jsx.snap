// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="w-full lg:w-3/5 flex-grow-1 relative px-4 lg:px-8 order-1"
>
  <div
    className="mx-auto lg:max-w-lg md:max-w-3xl py-4 lg:py-8 flex flex-col"
  >
    <div
      className="flex justify-center lg:justify-start"
    >
      <h3
        className="font-semibold !font-sans mb-8 tracking-tight text-custom-blue lg:text-gray-800 text-3xl order-first"
      >
        signUp
      </h3>
    </div>
    <div
      className="flex flex-col-reverse order-last lg:flex-col lg:order-2"
    >
      <button
        className="button__ButtonWrapper-sc-1aw98b4-0 button w-full h-14 px-6 transition"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "border": "2px solid black",
            "borderRadius": "0.5rem",
            "boxShadow": "2px 2px #1a1a1a",
          }
        }
      >
        <div
          className="relative flex h-full items-center space-x-4 justify-center"
        >
          <img
            alt="TwitterIcon"
            data-nimg="1"
            decoding="async"
            height={35}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style={
              Object {
                "color": "transparent",
              }
            }
            width={35}
          />
          <span
            className="block w-max font-medium tracking-wide text-gray-900 sm:text-base"
          >
            signWithTwitter
          </span>
        </div>
      </button>
      <button
        className="button__ButtonWrapper-sc-1aw98b4-0 button w-full h-14 px-6 transition"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "border": "2px solid black",
            "borderRadius": "0.5rem",
            "boxShadow": "2px 2px #1a1a1a",
          }
        }
      >
        <div
          className="relative flex h-full items-center space-x-4 justify-center"
        >
          <img
            alt="Google Icon"
            data-nimg="1"
            decoding="async"
            height={45}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
            srcSet="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
            style={
              Object {
                "color": "transparent",
              }
            }
            width={45}
          />
          <span
            className="block w-max font-medium tracking-wide text-gray-900 sm:text-base"
          >
            signWithGoogle
          </span>
        </div>
      </button>
      <div
        className="flex mt-7 order-3 items-center text-center mb-4"
      >
        <hr
          className=" text-gray-400  w-full "
        />
        <label
          className="block font-medium text-sm text-gray-700 ml-6 mr-6"
        >
          or
        </label>
        <hr
          className=" text-gray-400 w-full"
        />
      </div>
    </div>
    <form
      className="-mx-3 order-3"
      onSubmit={[Function]}
    >
      <div
        className="sm:flex flex-row hidden"
      >
        <div
          className="px-3 mb-3 w-full "
        >
          <div
            className="input__InputWrapper-sc-1hj52gj-0 hyTiyQ input-wrapper floating-label"
          >
            <input
              autoFocus={true}
              className="text-input !bg-white !border-2 !border-[black] !rounded-lg !shadow-none focus:[&:not(#password)]:!border-[3px] focus:[&:not(#password)]:!border-[black] focus:ring-0"
              id="name"
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              pattern={null}
              placeholder="name"
              required={true}
              type="text"
              value=""
            />
            <label
              className="form-label"
              htmlFor="name"
            >
              name
            </label>
          </div>
        </div>
        <div
          className="px-3 mb-3 w-full "
        >
          <div
            className="input__InputWrapper-sc-1hj52gj-0 hyTiyQ input-wrapper floating-label"
          >
            <input
              className="text-input !bg-white !border-2 !border-[black] !rounded-lg !shadow-none focus:[&:not(#password)]:!border-[3px] focus:[&:not(#password)]:!border-[black] focus:ring-0"
              id="surname"
              name="surname"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              pattern={null}
              placeholder="surname"
              required={true}
              type="text"
              value=""
            />
            <label
              className="form-label"
              htmlFor="surname"
            >
              surname
            </label>
          </div>
        </div>
      </div>
      <div
        className="px-3 mb-3 w-full "
      >
        <div
          className="input__InputWrapper-sc-1hj52gj-0 hyTiyQ input-wrapper floating-label"
        >
          <input
            className="text-input !bg-white !border-2 !border-[black] !rounded-lg !shadow-none focus:[&:not(#password)]:!border-[3px] focus:[&:not(#password)]:!border-[black] focus:ring-0"
            id="email"
            name="email"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            pattern={null}
            placeholder="email"
            required={true}
            type="email"
            value=""
          />
          <label
            className="form-label"
            htmlFor="email"
          >
            email
          </label>
        </div>
      </div>
      <div
        className="px-3 mb-3 w-full "
      >
        <div
          className="input__InputWrapper-sc-1hj52gj-0 hyTiyQ input-wrapper floating-label toggle-password"
        >
          <input
            className="text-input default !bg-white !border-2 !border-[black] !rounded-lg !shadow-none focus:[&:not(#password)]:!border-[3px] focus:[&:not(#password)]:!border-[black] focus:ring-0"
            id="password"
            name="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            pattern="^(?=.*[A-Z])(?=.*\\\\d)[A-Za-z\\\\d*&$@!%#?^(){}[\\\\]<>|~+=_-]{8,32}$"
            placeholder="password"
            required={true}
            type="password"
            value=""
          />
          <label
            className="form-label"
            htmlFor="password"
          >
            password
          </label>
          <button
            aria-label="Show/Hide password"
            className="show-password-button"
            onClick={[Function]}
            type="button"
          />
        </div>
        <p
          className="italic"
        >
          passwordCriteria
        </p>
      </div>
      <div
        className="pb-3 px-3"
      >
        <p>
          haveAccount
          <a
            href="/signin"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <span
              className="text-gray-800 underline"
            >
              signIn
            </span>
          </a>
        </p>
        <button
          className="button__ButtonWrapper-sc-1aw98b4-0 button !rounded-none !text-black-100 !bg-transparent !shadow-none !border-none after:content-[url(\\"/arrow.svg\\")] after:mt-[1px] ltr:after:ml-4 after:absolute rtl:after:rotate-180 rtl:after:mr-4"
          disabled={false}
          hidden={true}
        >
          forgetPassword
        </button>
      </div>
      <button
        className="w-full lg:w-auto bg-primary-200 focus:outline-none text-white font-semibold rounded-lg text-md px-5 py-3 flex gap-x-2 items-center"
      >
        signUp
        <span
          aria-hidden={true}
          className="loading__LoadingWrapper-sc-43dx6p-0 hEsmJL hidden size-sm style-orange spinner"
          size="sm"
        />
      </button>
    </form>
  </div>
</div>
`;
